libdnf/base/transaction.cpp

```diff
@@ -299,9 +299,12 @@ TransactionPackage Transaction::Impl::make_transaction_package(
 
     for (auto replaced_id : obs) {
         rpm::Package replaced_pkg(base, rpm::PackageId(replaced_id));
-        auto old_reson = replaced_pkg.get_reason();
-        if (old_reson > reason) {
-            reason = old_reson;
+        auto old_reason = replaced_pkg.get_reason();
+        if (old_reason == transaction::TransactionItemReason::NONE &&
+            reason <= transaction::TransactionItemReason::DEPENDENCY) {
+            reason = transaction::TransactionItemReason::NONE;
```

Comment:
The issue with your design is that it combines another boolean property, `Package::is_installed()` into a new value of the `Reason` enum.

Effectively, `pkg.get_reason() == INSTALLED_OUTSIDE_DNF` is equivalent to `pkg.is_installed && pkg.get_reason() == NONE`. I am convinced this alone is an obvious proof of a bad design. In general, if there's a need for such combination for convenience (or performance) reasons, the two elementary values should be left untouched and a third, compound value should be introduced that combines them into another semantically important attribute.

The `NONE` value as of now represents that there is no record of the user selecting a reason for a package in the system state toml. You are effectively splitting the value into two, with your new reason, if someone on the API would want to find out if a package has no reason, they will need the following condition: `pkg.get_reason() == NONE || pkg.get_reason() == INSTALLED_OUTSIDE_DNF`.

As for the argument of it being faster with not having to query the system repo every time you need to distinguish the two, instead there's the added complexity of the system state class needing to have the system repo as an input so that it can query, precompute and cache the information. Since it's effectively being a cache for this, there are also other ways to cache this without needing to add another enum value on the interface.

> ...
> I think that the solution is
>
>     * More transparent because the reason will explicitly says what it means (improvement to DNF4)
>
>     * System state will record INSTALLED_BY_DIFFERENT_TOOL as userinstalled=1 ...

No it won't, it'd still be missing a record for that NA, otherwise we'd erase the distinction and the reason would become USERINSTALLED. This is fine though and not an issue with the approach (AFAICS).

>     * ... but history DB will store INSTALLED_BY_DIFFERENT_TOOL

Normal history DB records won't have this. Only the "differential transactions" that we talked about (automatically created records of unexpected system state diffs we detect) could have it, but we could just as well use NONE for it.

>     * This will be less confusing for installonly packages because can be installed in multiple copies without removal

I don't understand this point.

>     * Better understandable outside of code of goal (repoquery) and in history DB

I don't think it makes a difference in history DB. NONE can represent the information just as well.

As for better understandable code, it depends on what information you need to retrieve. If you need to know if there's no reason record for a package, you'll now need to check for two enum values instead of one, which is worse.

>     * Performance - for available packages it will not require any calculation (looking into dictionary of system state, no RPM queries). It is remarkable because available packages represents 29/30 in our data set

I've addressed this above with the "cache" argument.

>     * Bulletproof solution - anytime when anyone will try to compare reasons - it will not require any additional code around to cover some cases. We already discussed that such a comparison is required 3 times in code and we already talked about additional usage for sorting of transaction elements.

I'd prefer you didn't use the word bulletproof in the context of dnf reasons... it's a house of cards one way or the other.

The need for the workarounds is because dnf has the convention of treating INSTALLED_OUTSIDE_DNF as userinstalled. That's the real cause of this issue.

I'm not even sure the comparator should be on the API, IIRC when we discussed it you said it's supposed to only be internal.

>     * Simplification - the code will be only represented by a single line - `reason = std::max(reason, old_reason);`

This simplification here comes at the cost of more complexity elsewhere.

It's not that I don't see the advantages of having the extra value, but I consider the benefits marginal, the complexity comparable (just in different places) and the design issues I outlined simply push the needle into the red area for me.

Reply:
It would be nice to see the issue with reasons in a bigger perspective. Or from history db view. When I see record in history DB with None reason I will suggest that it is bug in DNF. Also I can consider when I install packages as weak dependency than in the next upgrade transaction the  reason will change to dependency. Also the interesting part is when package will be installed as group and as a user but history structure does not support that complexity therefore we have another difference. And there is the last aspect, we have not much time left for development. It means that we have huge inconsistency between design of history DB and system state.